cosmos/runtime/txpool/mempool.go

```diff
@@ -125,28 +144,55 @@ func (m *Mempool) Insert(ctx context.Context, sdkTx sdk.Tx) error {
 		return errors.New("wraped tx is nil")
 	}
 
-	// If the tx is a local, or has been gossiped again for some reason. We ignore it.
-	ethTxHash := ethTx.Hash()
-	if m.txpool.Has(ethTxHash) {
+	// If we are currently protecting against block inserts, we queue the transaction
+	// to be inserted until after we are ready.
+	select {
+	case <-sCtx.Done():
+		return sCtx.Err()
+	case m.insertQueue <- ethTx:
 		return nil
 	}
+}
 
-	// Track time it entered from comet.
-	m.receivedFromCometAtMu.Lock()
-	m.receivedFromCometAt[ethTxHash] = time.Now()
-	m.receivedFromCometAtMu.Unlock()
-
-	if errs := m.txpool.Add(
-		[]*ethtypes.Transaction{wet.Unwrap()}, false, false,
-	); len(errs) != 0 {
-		// Handle case where a node broadcasts to itself, we don't want it to fail CheckTx.
-		if errors.Is(errs[0], ethtxpool.ErrAlreadyKnown) && sCtx.ExecMode() == sdk.ExecModeCheck {
-			return nil
+// processInserts processes inserts into the txpool.
+// this is basically used to process remote transactions
+// incoming from comet.
+func (m *Mempool) processInserts() {
+	m.wg.Add(1)
```

Comment:
nit: generally you call wg.Add from the main GR (the GR that blocks on wait), this is to avoid potential race-conditions between concurrent Add / Wait calls